Group key witness support

jharveyb opened the pull request:
Implements changes outlined in https://github.com/lightninglabs/taproot-assets/issues/466.

Known TODOs:

- [ ] VM execution in `DeriveGroupKey` to verify witness correctness
- [ ] Copy group witness to `asset.PrevWitnesses` where need so the witness gets serialized
- [ ] update `proof.Verify` to use new genesis asset definition
- [ ] Resolve dependency cycle with `tapscript` package
- [ ] New unit test cases for features introduced in past PRs

Approach:

- Modify the `virtualTx` methods to add new branches for creating a virtual TX for a genesis grouped asset.

- Modify the VM to prevent an early exit for genesis grouped assets, and add early assertions for such assets.

- Update `DeriveGroupKey` to sign over a virtualTx representing minting of a genesis grouped asset.

- Update `DeriveGroupKey` to accept external scripts and support different signing methods.

Open questions:

- Can we reuse the existing `VirtualTxSigner` for the `GenesisSigner`?

- Is there other data needed for a group key script path authorization here besides a single `TapLeaf`? In `CreateTaprootSignature` I see validation against other fields but not use of those other fields for signing.

- How can we resolve the `tapscript` dependency cycle? I see two issues, one being the placement of `RawKeyGenesisSigner` code, which AFAICT is only used in tests. The other is that, the `asset` package can't call methods from the `tapscript` package without causing a cycle, but with the current setup `DeriveGroupKey` must build a virtual TX at some point. I think one option may be to move some of the group signing logic into the `tapscript` package.

jharveyb commented:
I don't see a neat solution to the dependency cycle issues, so just moving some of the virtual TX creation logic for these special group genesis assets into `asset` and accepting some duplication with `tapscript` to keep this moving forward.

halseth commented on asset/asset.go:
```diff
@@ -967,6 +968,27 @@ func (a *Asset) HasGenesisWitness() bool {
 	return *witness.PrevID == ZeroPrevID
 }
 
+// HasGenesisWitnessForGroup determines whether an asset has a valid witness
```
remove use of word "valid" perhaps? Since we are not checking the signature

halseth commented on tapscript/tx.go:
```diff
@@ -71,65 +71,99 @@ func virtualTxInPrevOut(root mssmt.Node) *wire.OutPoint {
+		copyWithoutWitness.PrevWitnesses[i].TxWitness = nil
+	}
+
+	key := asset.ZeroPrevID.Hash()
```
Yeah, this looks correct I think! During VM execution the prev witness will be zero and point back to this zero-id.

halseth commented on tapscript/tx.go:
```diff
@@ -220,8 +254,19 @@ func virtualTxOut(asset *asset.Asset) (*wire.TxOut, error) {
+
 	// We'll start by mapping all inputs into a MS-SMT.
-	txIn, inputTree, err := virtualTxIn(newAsset, prevAssets)
+	switch newAsset.HasGenesisWitnessForGroup() {
```
hehe, just `if/else`? 😛

jharveyb replied:
I thought we don't do else 😅  Can change.

halseth commented on tapscript/tx.go:
```diff
@@ -220,8 +254,19 @@ func virtualTxOut(asset *asset.Asset) (*wire.TxOut, error) {
-	txIn, inputTree, err := virtualTxIn(newAsset, prevAssets)
+	switch newAsset.HasGenesisWitnessForGroup() {
+	case true:
+		txIn, inputTree, err = virtualGenesisTxIn(newAsset)
```
Very nice, diff was not as large as I expected 👍

halseth commented on tapscript/tx.go:
```diff
@@ -263,8 +308,19 @@ func InputAssetPrevOut(prevAsset asset.Asset) (*wire.TxOut, error) {
+		// If the input asset is a genesis asset that is part of an
+		// asset group, we need to validate the group witness against
+		// the tweaked group key and not the genesis asset script key.
+		if prevAsset.HasGenesisWitnessForGroup() {
```
My idea here was to set the `prevAsset.ScriptKey = groupKey` when creating the `prevAsset` copy. Is that feasible?

jharveyb replied:
Covered here:

https://github.com/Roasbeef/bips/pull/33#discussion_r1259002226

I think this still needs to get reflected in the BIP text. AFAICT, the `pkScript` here is what we actually pass to the internal `txscript.Engine` to validate the witness against, so it's the one we need to now be the group key. The asset used to build the virtual TX should have the tweaked group key set, but not swapped for the script key.

halseth replied:
See my comment linked on the BIP:

> The modification only affects the referenced previous leaf, that is used to construct the previous outpoint for the virtual tx. The output leaf (still committing to asset_script_key) remains the same.

jharveyb replied:
Resolved in BIP repo.

halseth reviewed (commented):
Def looks to be on the right track! 😀

jharveyb commented:
> What do you mean by this? The script path auth is just the existing witness field right? Or do you mean the group key derivation reveal? The group key witness (for minting) should be more or less the same as the flow for signing virtual txns normally. IIUC, we don't need any _new_ fields in the `VInput`, instead the existing fields are populated as normal. The main difference now is that we swap out the public key (not the input asset) for the group key asset. 

From the RPC level, what data types to I need to provide as arguments when minting a grouped asset into a group, where I want to use a script path and not a simple signature?

And then, how does that get passed in to the final signing descriptor (I think this part is clearer to me now).

Roasbeef commented:
> From the RPC level, what data types to I need to provide as arguments when minting a grouped asset into a group, where I want to use a script path and not a simple signature?

Ah, I think for that, we'd just make sure we're able to handle such spends internally via unit tests, then defer the expansion of the RPC interface to satisfy more complex group minting conditions. So to start, we can just allow them to specify their own group key and we use that instead of the auto generated one. The caller would then need to handle storing all the tapcript leaves, just like the itests we have to do non keyspend spends for normal assets. 

With that said, if we wanted to do some design sketches, I'd say: 
> what data types to I need to provide as arguments when minting a grouped asset into a group, where I want to use a script path and not a simple signature?

You'd need to specify:
  * The group key to use.
  * The full witness field for that group key. 

We can't necessarily generate the witness ourselves, as we may not even control the key used. Instead, the caller would use the `vPSBT` APIs to craft the minting transaction. To facilitate this, we'd need to expose calls that can be used to lock UTXOs (they want to know which outpoint will be used), or we'd add that as an arg to the minting process. 

> And then, how does that get passed in to the final signing descriptor (I think this part is clearer to me now).

We don't do signing in this context, as we might not be able to. The minting witness is just the witness of a "normal" spend, but it can spend the "all zero" outpoint, as it's a proof that the group key _can_ be satisfied. You can think of it like the PSBT based channel funding for lnd: we don't know what inputs the caller is using, but we can provide them with the information needed to make the funding transaction (the funding outpoint).

Roasbeef commented on vm/vm.go:
```diff
@@ -290,6 +299,21 @@ func (vm *Engine) Execute() error {
+		// For genesis assets in an asset group, set the previous asset
+		// as the genesis asset.
+		vm.prevAssets = commitment.InputSet{
+			asset.ZeroPrevID: vm.newAsset,
```
👍

Roasbeef commented on asset/asset.go:
```diff
@@ -469,10 +473,21 @@ func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+	return GroupPubKey(rawKey, assetID[:], g.TapscriptRoot[:])
+}
+
+func GroupPubKey(rawKey *btcec.PublicKey, singleTweak, tapTweak []byte) (
```
Missing `godoc`` comment.

Roasbeef commented on asset/asset.go:
```diff
@@ -727,14 +742,11 @@ func NewScriptKeyBip86(rawKey keychain.KeyDescriptor) ScriptKey {
-		*Genesis) (*btcec.PublicKey, *schnorr.Signature, error)
+	// SignVirtualTx generates a signature according to the passed signing
+	// descriptor and TX.
+	SignVirtualTx(signDesc *lndclient.SignDescriptor, tx *wire.MsgTx,
```
For our purposes, I think we can assume it's just keyspends for now, so we can actually keep the existing `KeyDescriptor`? Still getting thru the diff though.

Roasbeef replied:
Alternatively, we can rely on this method to produce the virtual tx itself. So it just needs the `Genesis` for the "root" asset, then the asset being created (`asset.Asset`), then it returns the witness as is. Also I don't think the first return value is valid any longer with the new scheme.

jharveyb replied:
Resolved offline.

Roasbeef commented on asset/asset.go:
```diff
@@ -757,42 +769,27 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
-	sig, err := schnorr.Sign(tweakedPrivKey, idHash[:])
+	// Derive the final tweaked public key also used for signing.
+	tweakedGroupKey, err := GroupPubKey(
+		signerPubKey, signDesc.SingleTweak, signDesc.TapTweak,
```
See above, I think we can keep the sign desc as is, then it just derives the tweaks from: the asset passed in, and the genesis to use for signing purposes. We can also tweak that slightly to be `GroupKey` or w/e (assuming it includes the key and the genesis information).

Roasbeef replied:
This way the interface is a bit more opinionated.

jharveyb replied:
Pretty sure this is addressed.

Roasbeef commented on asset/asset.go:
```diff
@@ -757,42 +769,27 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
-
-		id = currentGen.ID()
-	}
+	sig, err := SignVirtualTx(r.privKey, signDesc, virtualTx, prevOut)
```
Error is ignored here.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+		RawKey:        groupKey.RawKey,
+		GroupPubKey:   *tweakedGroupKey,
+		TapscriptRoot: groupKey.TapscriptRoot,
+		Witness:       witness,
```
Why do we need to attach a witness here? The witness would reside in the asset being created (the very first one), then we just need to care about the key and tapscript root. Then this can be re-used everywhere, with the group minting identified based on the asset itself (prev id zero, etc). 

Or seems this is both the initial derivation _and_ also the first signing? 

For group keys in assets, we don't actually need to pass that witness along each time for serialization, so I assume this is just for book keeping or convenience purpose?

jharveyb replied:
In the minter, we set the group key info of the asset based on the output of `DeriveGroupKey`. I think ideally we'd modify `asset.New()` to copy the witness from here passed `GroupKey` into the genesis asset witness for the new asset.

jharveyb replied:
The actual group witness copy into the genesis asset witness is here in `asset.New()`:

https://github.com/lightninglabs/taproot-assets/blob/843584a12a7bcefd85003f584c24fe48e8705713/asset/asset.go#L951

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
-func DeriveGroupKey(genSigner GenesisSigner, rawKey keychain.KeyDescriptor,
-	initialGen Genesis, currentGen *Genesis) (*GroupKey, error) {
+func DeriveGroupKey(genSigner GenesisSigner, groupKey *GroupKey,
+	scriptLeaf *psbt.TaprootTapLeafScript, initialGen Genesis,
```
I think we can use the `AssetGroup` struct in lieu of the group key, then the initial genesis?

jharveyb replied:
Yeah the `GroupKey` was intended to carry an optional tapscript root, but if that's no longer needed then `AssetGroup` does make more sense.

jharveyb replied:
GroupKey was dropped so i think the `DeriveGroupKey` args are ok as-is.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
-func DeriveGroupKey(genSigner GenesisSigner, rawKey keychain.KeyDescriptor,
-	initialGen Genesis, currentGen *Genesis) (*GroupKey, error) {
+func DeriveGroupKey(genSigner GenesisSigner, groupKey *GroupKey,
+	scriptLeaf *psbt.TaprootTapLeafScript, initialGen Genesis,
```
Also why `scriptLeaf` vs the actual script root? We may not know what the leaves are (for now caller will handle the external witness signing, etc).

jharveyb replied:
Moved out of scope for this PR.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+		return nil, fmt.Errorf("asset group type mismatch")
+	}
+
+	if scriptLeaf != nil && groupKey.TapscriptRoot == [32]byte{} {
```
At this phase, I don't think we can really assert much about the script leaf. 

For now we can handle just the keyspend path. Anything more advanced than that (musig2, etc, etc) will require the caller to drive the API to do the fancy stuff.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+		SingleTweak:   genesisTweak[:],
+		DoubleTweak:   nil,
+		TapTweak:      []byte{},
+		WitnessScript: []byte{},
```
If the script leaf was needed (don't think it is), then the actual script would be placed here.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+		InputIndex:    0,
+	}
+	if groupKey.TapscriptRoot != [32]byte{} {
+		signDesc.TapTweak = groupKey.TapscriptRoot[:]
```
So this is only actually needed when doing a key path spend.

jharveyb replied:
So sounds like for now we're only going to support signing for group keys with empty tapscript roots?

Roasbeef commented on asset/asset.go:
```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+		return nil, fmt.Errorf("cannot tweak group key: %w", err)
+	}
+
+	// Reimplement tapscript.InputAssetPrevOut to avoid a dependency cycle.
```
Leaving comment here so we make sure it's addressed. Alternatively, we can just move this function out of the `asset` package and into `tapscript`?

jharveyb commented on asset/asset.go:
```diff
@@ -967,6 +968,27 @@ func (a *Asset) HasGenesisWitness() bool {
+// key, and a single prevWitness with a zero PrevID, empty split commitment
+// proof, and non-empty witness.
+func (a *Asset) HasGenesisWitnessForGroup() bool {
+	if a.GroupKey == nil || len(a.PrevWitnesses) != 1 {
```
I think this still needs some change, we actually have two new cases:

-An asset with the existing genesis asset witness, that also has the group key set (needs a witness generated).
-An grouped genesis asset with a witness (empty PrevID, nil SplitCommitment, non-nil TxWitness, non-nil GroupKey), should still be considered a genesis asset

jharveyb replied:
This has been addressed with two new methods, `NeedsGenesisWitnessForGroup` and `HasGenesisWitnessForGroup`.

The `Needs` variant covers an asset with a genesis asset witness, and a group key set. This is used as a sanity check within the `GenesisTxBuilder` implementation, and to branch on in the virtual TX construction logic.

The `Has` variant is used in the VM and the proof verification logic, since at those points any genesis grouped asset should have its group witness attached, and now it only needs to be validated. At those points an asset with an empty group witness would be considered invalid.

jharveyb commented on tapscript/tx.go:
```diff
@@ -71,65 +71,99 @@ func virtualTxInPrevOut(root mssmt.Node) *wire.OutPoint {
-	// an empty input tree.
-	isGenesisAsset := newAsset.HasGenesisWitness()
 	inputTree := mssmt.NewCompactedTree(mssmt.NewDefaultStore())
-	if !isGenesisAsset {
```
Yes, this point was never reached with a genesis asset before because the VM would have exited early on any genesis assets before this line.

Now, we have `virtualGenesisTxIn` and a genesis asset should still never appear at this point, so the branch is unneeded.

jharveyb commented:
Read through the feedback, my current plan is:

- Move the virtual TX construction logic behind another interface, `asset.GenesisTxBuilder`, to resolve dependency issues.
- Simplify `DeriveGroupKey` to only handle the key spend path with an empty tapscript root.
- New methods on `Asset` so its clearer when a grouped genesis asset needs a witness or already has a witness.
- Update `asset.new()` or add logic to the minter s.t. the group key witness is copied into the asset during minting.

jharveyb commented:
Latest version is on this branch with the described changes. The commits were re-done and should be relatively readable.

Moving on to improving test coverage.

jharveyb commented:
Another source of dependency issues:

https://github.com/lightninglabs/taproot-assets/blob/843584a12a7bcefd85003f584c24fe48e8705713/commitment/asset.go#L151

Verifying the group witness on asset commitment insertion.

Roasbeef commented:
> Verifying the group witness on asset commitment insertion.

So I think we can actually move this out entirely. Looks like the only time we call `VerifySignature` is `Upsert` and also when we make a new asset commitment. Now that we have the group key reveal stuff in the proof file verification, we can just move that there. So now we only verify the signature when we create assets, and also when we verify proofs for assets.

Roasbeef commented on asset/mock.go:
```diff
@@ -61,6 +65,160 @@ func RandGroupKeyWithSigner(t testing.TB, genesis Genesis) (*GroupKey, []byte) {
+
+// Forked from tapscript/tx/virtualTxIn to remove checks for non-nil PrevIDs on
+// inputs and full consumption of inputs.
+func VirtualGenesisTxIn(newAsset *Asset) (*wire.TxIn, mssmt.Tree, error) {
```
I think we can move this out of `mock.go`, so residing in like `asset/group_genesis.go` or something like that. After we get things all working we maybe just want to move it purely within `txscript`, then we keep the `asset` package concerned with mainly serialization and initial sanity checks.

jharveyb replied:
Moved in latest version out to `asset/asset` near the other group logic - I agree that we probably want to rearrange these implementations overall.

Roasbeef commented on tapscript/mint.go:
```diff
@@ -0,0 +1,50 @@
+
+// MockGroupTxBuilder is an implementation of the asset.GenesisTxBuilder
+// interface that constructs virtual transactions for grouped asset genesis.
+type BaseGroupTxBuilder struct{}
```
godoc name mismatch?

Roasbeef commented on asset/asset.go:
```diff
@@ -756,47 +743,61 @@ func NewRawKeyGenesisSigner(priv *btcec.PrivateKey) *RawKeyGenesisSigner {
+type RawGroupTxBuilder struct{}
 
-		id = currentGen.ID()
+func (m *RawGroupTxBuilder) BuildGenesisTx(newAsset *Asset) (*wire.MsgTx,
```
Does this version need to live in the `asset` package? Can we get away with just the version in `tapscript`?

jharveyb replied:
Not really, since that would cause dependency cycles for many tests that use `RandGroupKey()`. Similar to how we had the raw key genesis signer and the real `gen_signer`.

Roasbeef commented on tapscript/mint.go:
```diff
@@ -0,0 +1,50 @@
+// interface that constructs virtual transactions for grouped asset genesis.
+type BaseGroupTxBuilder struct{}
+
+func BuildGenesisTx(newAsset *asset.Asset) (*wire.MsgTx,
```
Missing a `godoc` template.

Roasbeef commented on virtual_tx_builder.go:
```diff
@@ -0,0 +1,22 @@
+// that constructs virtual transactions for grouped asset genesis.
+type GroupTxBuilder struct{}
+
+func (b *GroupTxBuilder) BuildGenesisTx(newAsset *asset.Asset) (*wire.MsgTx,
```
Looks like the 3rd instance of this in the diff? Wondering if we can consolidate somewhat. Might need to modify the structure of the mock packages and/or tests.

Roasbeef commented on vm/vm.go:
```diff
@@ -191,15 +191,24 @@ func (vm *Engine) validateWitnessV0(virtualTx *wire.MsgTx, inputIdx uint32,
-	if err != nil {
-		return err
+	// Genesis grouped assets will have a nil PrevID and match the prevAsset
+	// since it is a copy of the original asset.
```
Maybe rewrite comment like: 
> For assets w/o a group witness, we want to check the prev asset to make sure it's spending like-for-like.

Roasbeef commented on asset/asset.go:
```diff
@@ -468,10 +469,25 @@ func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
 	}
 
-	internalKey := input.TweakPubKeyWithTweak(rawKey, assetID[:])
+	if g.TapscriptRoot != nil {
```
Can we just collapse this? So just calling it with nil directly.

Roasbeef commented on asset/asset.go:
```diff
@@ -468,10 +469,25 @@ func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+		return GroupPubKey(rawKey, assetID[:], g.TapscriptRoot[:])
+	}
+
+	return GroupPubKey(rawKey, assetID[:], nil)
```
`[]byte{}`?

Roasbeef commented on asset/asset.go:
```diff
@@ -468,10 +469,25 @@ func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+func GroupPubKey(rawKey *btcec.PublicKey, singleTweak, tapTweak []byte) (
+	*btcec.PublicKey, error) {
+
+	if len(singleTweak) != sha256.Size || len(tapTweak) != sha256.Size {
```
`tapTweak` can be `nil`, if there's no tapscript root.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,19 +817,58 @@ var _ GenesisTxBuilder = (*RawGroupTxBuilder)(nil)
+	// Compute the tweaked group key and set it in the asset before
+	// creating the virtual minting transaction.
+	genesisTweak := initialGen.ID()
+	tweakedGroupKey, err := GroupPubKey(rawKey.PubKey, genesisTweak[:], nil)
```
Eventually should also have the script root here. Could have it be a functional option or something like that for now.

Roasbeef commented on asset/asset.go:
```diff
@@ -801,19 +817,58 @@ var _ GenesisTxBuilder = (*RawGroupTxBuilder)(nil)
+	// tapscript root and key path spends.
+	signDesc := &lndclient.SignDescriptor{
+		KeyDesc:     rawKey,
+		SingleTweak: genesisTweak[:],
```
Don't have the double tweak rn as the tapscript layer isn't there?

Roasbeef commented on asset/asset.go:
```diff
@@ -939,18 +939,26 @@ func New(genesis Genesis, amount, locktime, relativeLocktime uint64,
+
+	// Genesis assets with an asset group must have the group witness stored
+	// in the genesis asset witness, if present.
+	if groupKey != nil && groupKey.Witness != nil {
```
Still think we don't necessarily need the witness field within the `GroupKey` itself. I can see how it's convenient here though since we're already accepting in the group key.

guggero replied:
Also, should we use `len(groupKey.Witness) != 0` to avoid this being true for empty witnesses?

Roasbeef replied:
Should use length here, otherwise this should get rejected later as it has a witness, but may not be prev out zero or nl,e tc.

jharveyb commented:
Addressed a few comments, updated this main branch and the WIP branch with unfinished test fixes here:

https://github.com/lightninglabs/taproot-assets/tree/group-key-witness-support-wip

There was an issue with not using the correct branching condition in `VirtualTx` for reaching `VirtualGenesisTxIn` - since `VirtualTx` is called both when building the Virtual TX for signing, and when verifying in the VM, I think we'd actually want to branch on the genesis grouped asset both before and after the group witness has been attached.

However, VM validation for normal state transitions is now failing, so I think the branch condition as I've written it is too broad - I think this could be fixed by also considering `prevAssets` at that point. So something like:

If both the new asset and the prev asset are genesis grouped assets, then it must be a genesis grouped asset mint, branch to VirtualGenesisTxIn.

We still have the sanity check very early in the VM that I think should stay as `HasGenesisWitnessForGroup`.

The WIP branch also has some fixes for other tests but those haven't been exercised yet since they usually depend on the VM working anyways.

guggero commented on tapgarden/caretaker.go:
```diff
@@ -536,6 +536,15 @@ func (b *BatchCaretaker) seedlingsToAssetSprouts(ctx context.Context,
+		if sproutGroupKey != nil {
+			err := b.cfg.TxValidator.Execute(newAsset, nil, nil)
+			if err != nil {
+				return nil, fmt.Errorf("unable to verify"+
```
nit: missing space after "verify".

Roasbeef replied:
Log will be mangled without the space.

guggero replied:
Still missing in latest diff.

guggero commented on commitment/commitment_test.go:
```diff
@@ -342,28 +332,41 @@ func TestMintTapCommitment(t *testing.T) {
+			invalidCollctibleAmt := amt != nil && *amt != uint64(1) &&
+				details.Type == asset.Collectible
+
+			switch invalidNormalAmt || invalidCollctibleAmt {
```
nit: same here re true/false switch statements.

guggero commented on commitment/commitment_test.go:
```diff
@@ -342,28 +332,41 @@ func TestMintTapCommitment(t *testing.T) {
+			// group key creation.
+			var err error
+			amt := details.Amount
+			invalidNormalAmt := details.Type != asset.Collectible &&
```
We could create two package-level variables `zero` and `one` to make it easier to use with the collectible amounts.

guggero commented on vm/vm.go:
```diff
@@ -191,24 +191,37 @@ func (vm *Engine) validateWitnessV0(virtualTx *wire.MsgTx, inputIdx uint32,
+		}
+
+		// The parameters of the new and old asset much match exactly.
+		err := matchesAssetParams(vm.newAsset, prevAsset, witness)
```
This overwrites the outer `err`, making the comparison below incorrect (this is what the linter is complaining about).

guggero reviewed (commented):
Great work, very nice to see this getting close.

jharveyb commented on asset/mock.go:
```diff
@@ -61,6 +63,87 @@ func RandGroupKeyWithSigner(t testing.TB, genesis Genesis) (*GroupKey, []byte) {
 }
+
+// Copied from tapscript/tx.go.
+func virtualGenesisTxWithInput(virtualTx *wire.MsgTx, input *Asset,
```
As in move them from tapscript and get rid of the redunant mock ones? I'm in favor.

jharveyb commented on tapscript/mint.go:
```diff
@@ -0,0 +1,50 @@
+	// First, we check that the passed asset is a genesis grouped asset
+	// that has no group witness.
+	if !newAsset.NeedsGenesisWitnessForGroup() {
+		return nil, nil, fmt.Errorf("asset is not a genesis grouped" +
```
Mm if we move the required TX utilities to asse to think we could have one implementation there and then the top-level wrapper, and get rid of the wrapper in the tapscript package.

jharveyb replied:
Turns out unifying these is the same as moving all the TX-related logic into `asset` - which I don't think we should do in this PR.

jharveyb commented on commitment/commitment_test.go:
```diff
@@ -329,8 +342,28 @@ func TestMintTapCommitment(t *testing.T) {
+				amt = *details.Amount
+			}
+
+			// TOO(jhb): fix nil amount issues, use this err
```
I consider it resolved by the logic below or skipping the group key logic but forgot to remove the TODO itself.

Roasbeef replied:
Can be removed now?

jharveyb commented:
Latest push is further fixes to address some unit test fixes + fix group insertion when syncing from a Universe (h/t guggero).

Most itests pass, but reissuance does not work - adding a `GroupVerifier` interface now to be used when verifying proofs, to check if there is a verified group that matches the group key in the proof.

jharveyb commented:
Adding a `groupVerifier` has worked for fixing the reissuance test.

From offline discussion, I realized that the GroupKeyReveal should be omitted for genesis proofs of reissued assets, and instead the asset's group key should only be checked against stored group keys to ensure that the daemon has previously verified the group anchor (along with group witness verification).

However, this introduces new ordering constraints in many places where we handle grouped assets (storing to disk, verifying a proof, or populating a universe tree). The anchor asset for a group MUST be inserted before attempting to verify or store any other member of that asset group.

Right now many DB operations are unordered over e.x. the set of all assets in a minting batch, which I'm updating now.

jharveyb commented:
All itests are now passing locally & with the current state of this PR. Latest changes were to sort assets such that group anchors are inserted on disk before any possible group reissuances, since group key verification now relies on the group anchor being present on disk.

Pending changes are addressing reviewer feedback and changing the `GroupKey.TapscriptRoot` member from `[32]byte` to `[]byte`. As an array the nil-ness of the field is harder to test, since it is set to [32]byte{0x0} when not explicitly initialized.

dstadulis commented:
TODOs:

- [ ] Adding feedback from previous review
- [ ] Array to slice changes

jharveyb commented on tapcfg/server.go:
```diff
@@ -278,13 +278,13 @@ func genServerConfig(cfg *Config, cfgLogger btclog.Logger,
+				Log:                   assetMintingStore,
+				KeyRing:               keyRing,
+				GenSigner:             virtualTxSigner,
+				GenTxBuilder:          &tap.GroupTxBuilder{},
```
Would be useful but I don't want to add that for this PR.

jharveyb commented on asset/tx.go:
```diff
@@ -109,3 +112,79 @@ func InputGenesisAssetPrevOut(prevAsset Asset) (*wire.TxOut, error) {
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot tweak group key: %w", err)
+	}
+	populatedVirtualTx := virtualGenesisTxWithInput(
```
Yes, though resolved now by moving some TX utilities to `asset`.

Roasbeef commented on asset/asset.go:
```diff
@@ -469,11 +469,30 @@ func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+	// A nil Tapscript root will be represented as an empty byte slice after
+	// encoding and decoding, so we need to check for both forms.
+	internalKey := input.TweakPubKeyWithTweak(rawKey, singleTweak)
+	if tapTweak == nil || bytes.Equal(tapTweak, []byte{}) {
```
Yeah both mean the same thing now that we're just using `[]byte` here.

Roasbeef commented on asset/asset.go:
```diff
@@ -726,19 +745,63 @@ func NewScriptKeyBip86(rawKey keychain.KeyDescriptor) ScriptKey {
-	initialGen Genesis, currentGen *Genesis) (*GroupKey, error) {
+func DeriveGroupKey(genSigner GenesisSigner, genBuilder GenesisTxBuilder,
+	rawKey keychain.KeyDescriptor, initialGen Genesis,
+	newAsset *Asset) (*GroupKey, error) {
```
I think we only need the `newAsset` argument as the genesis is now always the same (the one used to derive the key in the first place), and the asset include a `Genesis` instance.

jharveyb replied:
Not sure what you mean, `initialGen == newAsset.Genesis` only for the group creation. We also use the full group anchor genesis for an asset type check here instead of just passing the single tweak.

Roasbeef commented on asset/mock.go:
```diff
@@ -266,15 +261,29 @@ func RandID(t testing.TB) ID {
 	return a
 }
 
+// AssetNoErr creates an asset and fails the test if asset creation fails.
```
`NewAssetNoErr`?

Roasbeef commented on commitment/commitment_test.go:
```diff
@@ -329,8 +342,28 @@ func TestMintTapCommitment(t *testing.T) {
+			// to Mint() below, not on asset creation. Use the safe
+			// amount of 1 here, and pass the original amount to
+			// Min() below.
+			var amt uint64
```
Why wasn't this the case before?

jharveyb replied:
Should be resolved - we just want invalid amounts to fail because of a call to `Mint()` and not another function used for test setup.

Roasbeef commented on commitment/commitment_test.go:
```diff
@@ -352,15 +385,26 @@ func TestMintAndDeriveTapCommitment(t *testing.T) {
 
 	const assetType = asset.Normal
 	const numAssets = 5
+	var genesis1protoasset *asset.Asset
```
Style not: not fully camelCase, also can nest under the same `var` statement.

Roasbeef replied:
Also doesn't look like this one actually needs to be pre-declared?

Roasbeef commented on commitment/commitment_test.go:
```diff
@@ -422,7 +466,12 @@ func TestMintAndDeriveTapCommitment(t *testing.T) {
 	// with a distinct genesis and group key.
 	genesis2 := asset.RandGenesis(t, assetType)
-	groupKey2 := asset.RandGroupKey(t, genesis2)
+	genesis2protoasset := asset.AssetNoErr(
```
Same comment here re camel case.

Roasbeef commented on commitment/commitment_test.go:
```diff
@@ -527,9 +576,7 @@ func TestSplitCommitment(t *testing.T) {
 		{
 			name: "single input split commitment",
 			f: func() (*asset.Asset, *SplitLocator, []*SplitLocator) {
-				input := randAsset(
```
Hmm, perhaps we still want to retain the randomization here w.r.t the actual asset details? IIUC the new dep here is that for everything to validate, we need to create thing in the following order: genesis, group key, asset.

jharveyb replied:
Yup I think this is a leftover from debugging before we removed witness validation from commitment tree construction.

Roasbeef commented on commitment/commitment_test.go:
```diff
@@ -438,8 +487,16 @@ func TestSplitCommitment(t *testing.T) {
+	collectibleProtoAsset := randAsset(t, genesisCollectible, nil)
+	groupKeyNormal := asset.RandGroupKey(t, genesisNormal, normalProtoAsset)
+	groupKeyCollectible := asset.RandGroupKey(
+		t, genesisCollectible, collectibleProtoAsset,
```
Why's this need both the genesis and the actual asset? The asset includes the genesis.

Roasbeef replied:
Ah nvm, with the way that `DeriveGroupKey` func works, it always wants to sign, and during reissuance, you're signing the new asset.

Roasbeef commented on tapdb/universe.go:
```diff
@@ -232,9 +233,29 @@ func upsertAssetGen(ctx context.Context, db UpsertAssetStore,
-		_, err := upsertGroupKey(
-			ctx, groupKey, db, genPointID, genAssetID,
+	if groupKey != nil && genesisProof.GroupKeyReveal == nil {
+		fmt.Printf("nil group reveal!")
```
Stray print statement. Should be an error?

Roasbeef commented on tapdb/universe_test.go:
```diff
@@ -116,10 +116,21 @@ func randMintingLeaf(t *testing.T, assetGen asset.Genesis,
 		GenesisProof: randProof(t),
 		Amt:          uint64(rand.Int31()),
 	}
+	// The asset within the genesis proof is random; reset the asset genesis
```
Style nit: missing newline above.

Roasbeef commented on tapdb/universe_test.go:
```diff
@@ -460,6 +472,20 @@ func TestUniverseLeafQuery(t *testing.T) {
 		leaf := randMintingLeaf(t, assetGen, id.GroupKey)
+		if id.GroupKey != nil {
+			// All assets are sharing the same genesis and group
+			// key, so they must also shae the same group witness.
```
shae -> share

Roasbeef replied:
Is this for the case of a multi-mint, so same asset just with diff UTXOs? If not, then totally possilbe (if they're reissuance), that these values are the same, but they have a unique witness.

jharveyb replied:
It shouldn't be possible in practice because geneses can never be the same, even in the same batch (asset tags must be unique per-batch).

I think the closest real-world case would be a multi-asset group in the same batch, but for this test we randomly choose to use a group key for the universe ID, so I decided to match that randomization vs. forcing a group key.

Roasbeef commented on proof/mock.go:
```diff
@@ -49,6 +50,29 @@ func MockHeaderVerifier(header wire.BlockHeader, height uint32) error {
+// functions that require a group anchor verifier. This function is used in
+// those cases.
+func MockGroupAnchorVerifier(gen *asset.Genesis,
+	groupKey *asset.GroupKey) error {
```
I think this would also need the tapscript root to verify properly? Still getting through commit.

Roasbeef commented on proof/verifier.go:
```diff
@@ -356,6 +356,17 @@ func (p *Proof) verifyGroupKeyReveal() error {
+// GroupVerifier is a callback function which returns an error if the given
+// group key has not been imported by the tapd daemon. This can occur if the
+// issuance proof for the group anchor has not been imported or synced.
+type GroupVerifier func(groupKey *btcec.PublicKey) error
```
Do we just actually want a predicate here, or is the error useful down the line for test assertions?

jharveyb replied:
Error was useful while working on tests - I think also we want to pass any DB error that could occur inside the verifier vs. only handling the expected error of a missing group.

Roasbeef commented on proof/verifier.go:
```diff
@@ -356,6 +356,17 @@ func (p *Proof) verifyGroupKeyReveal() error {
+// GroupAnchorVerifier is a callback function which returns an error if the
+// given genesis is not the asset genesis of the group anchor. This would be
+// true for an reissuance into an existing group.
+type GroupAnchorVerifier func(gen *asset.Genesis,
```
Having trouble parsing the use of this from the function signature and also comment name. IIUC, you also need the tapscript root to verify the mapping, but then this seems like just the normal genesis reveal verification?

jharveyb replied:
Similar to the group key reveal verification, but in that context we're checking across different parts of the proof, and here we're checking against the DB or re-deriving the key. Made the comment more verbose.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,104 @@ func GenHeaderVerifier(ctx context.Context,
+
+		// This query will err if no stored group has a matching
+		// tweaked group key.
+		_, err := mintingStore.FetchGroupByGroupKey(
```
Can add a cache here in the future to speed things up.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,104 @@ func GenHeaderVerifier(ctx context.Context,
+	*asset.GroupKey) error {
+
+	// Cache anchors for groups that were previously fetched.
+	groupAnchors := make(map[asset.SerializedKey]*asset.Genesis)
```
`cache.LRU`? https://pkg.go.dev/github.com/lightninglabs/neutrino/cache@v1.1.2/lru#NewCache

Can also use above.

jharveyb replied:
Fixed size seems like a problem? We create these callbacks on startup and the size isn't bounded.

jharveyb replied:
Seems like something that could be added to `fn` with an Entry API similar to Rust HashMaps if we end up using this pattern a lot.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,104 @@ func GenHeaderVerifier(ctx context.Context,
+			groupAnchor = storedGroup.Genesis
+		}
+
+		if gen.ID() != groupAnchor.ID() {
```
Ah ok, so this verifies that the given asset is actually the root/anchor asset for the group key. In other words, we can use this to verify if an asset is the first asset to be issued with a group key.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,104 @@ func GenHeaderVerifier(ctx context.Context,
+			// TODO(jhb): add tapscript root support
+			singleTweak := gen.ID()
+			tweakedGroupKey, err := asset.GroupPubKey(
+				groupKey.RawKey.PubKey, singleTweak[:], nil,
```
Assumes no tapscript root is present.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,104 @@ func GenHeaderVerifier(ctx context.Context,
+	*asset.GroupKey) error {
+
+	// Cache group anchors we already verified.
+	groupAnchors := make(map[asset.SerializedKey]*asset.Genesis)
```
Can also use LRU cache linked above here with minimal overhead.

Roasbeef commented on proof/mint.go:
```diff
@@ -259,21 +263,28 @@ func committedProofs(baseProof *Proof, taprootAssetRoot *commitment.TapCommitmen
+			err := anchorVerifier(
+				&newAsset.Genesis, groupKey,
+			)
+			if err == nil {
```
I think adjusting the naming of `anchorVerifier` and also the callback type could make this easier to understand. 

IIUC, we're just checking here that this asset is the actual anchor asset for the group.

Roasbeef commented on proof/verifier.go:
```diff
@@ -453,28 +471,41 @@ func (p *Proof) Verify(ctx context.Context, prev *AssetSnapshot,
+	// 7. Verify group key for asset transfers. Any asset with a group key
+	// must carry a group key that has already been imported and verified.
+	if !isGenesisAsset && hasGroupKey {
+		if err := p.verfyGenesisGroupKey(groupVerifier); err != nil {
```
👍

Roasbeef commented on tapdb/asset_minting.go:
```diff
@@ -939,14 +939,27 @@ func (a *AssetMintingStore) AddSproutsToBatch(ctx context.Context,
+	// anchor verification depends on inserting group anchors before
+	// reissuances here. We use the raw group anchor verifier since there
+	// is not yet any stored asset group to reference in the verifier.
+	anchorVerifier := tapgarden.GenRawGroupAnchorVerifier(ctx)
```
Hmm couldn't we just rely on the upsert behavior here for the genesis?

jharveyb replied:
No -  the problem is that from the P.O.V. of the DB queries, the group anchor is defined as the grouped asset with a witness that was stored first.

This wasn't a problem before as only one group member, the group anchor, had a witness.

Now that all group members have a witness, their insertion must be ordered and the anchor must be inserted first. Otherwise the wrong genesis is returned when querying for the group info later on.

Roasbeef commented on address/address.go:
```diff
@@ -110,12 +109,6 @@ type Tap struct {
-	// membership of an asset in a partiular asset group. The witness can
-	// be a single signature or a script from the tapscript tree committed
-	// to with the TapscriptRoot, and follows the witness rules in BIP-341.
-	groupWitness wire.TxWitness
```
Hmm, not sure why we ever had this....

jharveyb replied:
Me neither, but I'm guessing it was related to the check in asset commitments.

jharveyb commented on asset/tx.go:
```diff
@@ -109,3 +112,79 @@ func InputGenesisAssetPrevOut(prevAsset Asset) (*wire.TxOut, error) {
+// GenesisSigner interface.
+var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+
+type RawGroupTxBuilder struct{}
```
Needed for tests where we can't import from `tapscript`.

jharveyb commented:
Updated to address reviewer feedback + BIP vectors are updated to add the new proof reveal fields + `GroupKey.TapscriptRoot` is now `[]byte`, which simplified handling in a few places.

guggero commented on tapdb/universe.go:
```diff
@@ -233,27 +233,36 @@ func upsertAssetGen(ctx context.Context, db UpsertAssetStore,
+	if groupKey != nil {
+		// Every group-related issuance must be accompanied by a group
+		// witness.
+		groupWitness := genesisProof.Asset.PrevWitnesses[0].TxWitness
```
Should we Squash the changes in this file with 42f097ac to make the review flow somewhat easier? Since this isn't really related to storing the group anchors early, just storing them correctly?

guggero commented on tapdb/universe.go:
```diff
@@ -233,27 +233,36 @@ func upsertAssetGen(ctx context.Context, db UpsertAssetStore,
+			fullGroupKey.RawKey = keychain.KeyDescriptor{
+				PubKey: rawKey,
+			}
+			if len(reveal.TapscriptRoot) != 0 {
```
I don't think we need a check or a copy here. Can just do `fullGroupKey.TapscriptRoot = reveal.TapscriptRoot` or not?

guggero commented on tapgarden/caretaker.go:
```diff
@@ -462,10 +473,23 @@ func (b *BatchCaretaker) seedlingsToAssetSprouts(ctx context.Context,
+				assetGen, amount, 0, 0, tweakedScriptKey, nil,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create"+
```
nit: space missing in error message before line break.

guggero commented on tapgarden/caretaker.go:
```diff
@@ -1188,3 +1189,113 @@ func GenHeaderVerifier(ctx context.Context,
+			return fmt.Errorf("cannot verify empty group key")
+		}
+
+		assetGroupKey := asset.ToSerialized(groupKey)
```
Since we don't know if this is accessed by goroutines in parallel, do we need to add a mutex to guard the cache as well?

jharveyb replied:
Good point, in the global config I made a separate once for each subsystem, but I forgot that some of these paralellize internally.

guggero commented on proof/append_test.go:
```diff
@@ -23,7 +23,7 @@ func genTaprootKeySpend(t testing.TB, privKey btcec.PrivateKey,
 	t.Helper()
 
-	virtualTxCopy := tapscript.VirtualTxWithInput(
+	virtualTxCopy := asset.VirtualTxWithInput(
```
nit: looks like this should be in an earlier commit to allow all commits to compile on their own (can use `git rebase -i v0-3-0-staging --exec "make unit pkg=... case=_none_"` to test that, though it looks like the `v0-3-0staging` branch has a compilation issue already, so no priority, can do that before merging it into `main`)

guggero commented on proof/mint.go:
```diff
@@ -259,21 +263,22 @@ func committedProofs(baseProof *Proof, taprootAssetRoot *commitment.TapCommitmen
+			err := anchorVerifier(
+				&newAsset.Genesis, groupKey,
+			)
+			if err == nil {
```
So what if the `err` _is not_ `nil`? Should we add a return and then do the below non-indented?

jharveyb replied:
We don't want to return nor skip the rest of this logic, since we still need to store the proof.

guggero commented on tapgarden/caretaker.go:
```diff
@@ -1164,6 +1185,39 @@ func SortSeedlings(seedlings []*Seedling) []string {
+			switch {
+			case err == nil:
+				anchorAssets = append(anchorAssets, fullAsset)
+			default:
```
Should we instead return any errors that occur here? Otherwise they might be swallowed and only surface at a later place where they might be harder to pinpoint.

jharveyb replied:
Added concrete errors so we can now pass along anything else.

guggero commented on tapdb/universe.go:
```diff
@@ -252,7 +252,7 @@ func upsertAssetGen(ctx context.Context, db UpsertAssetStore,
 		}
 		if len(reveal.TapscriptRoot) != 0 {
-			copy(fullGroupKey.TapscriptRoot[:], reveal.TapscriptRoot)
+			copy(fullGroupKey.TapscriptRoot, reveal.TapscriptRoot)
```
I don't think we need the `id` and the `copy` anymore, can just directly assign.

guggero commented on tapgarden/caretaker.go:
```diff
@@ -1164,6 +1185,39 @@ func SortSeedlings(seedlings []*Seedling) []string {
+	anchorVerifier proof.GroupAnchorVerifier) ([]*asset.Asset,
+	[]*asset.Asset) {
+
+	var anchorAssets, nonAnchorAssets []*asset.Asset
```
Since we expect the non-anchor witnesses to be the vast majority, should we go ahead and pre-initialize the slice to the size of `fullAssest`? Since if there are many thousands of assets, the time spent on growing the slice might not be insignificant.

guggero commented on universe/base.go:
```diff
@@ -264,26 +268,70 @@ func (a *MintingArchive) RegisterNewIssuanceBatch(ctx context.Context,
+	// verified and stored before any issuances that may be reissuances into
+	// the same asset group. This is required for proper verification of
+	// reissuances, which may be in this batch.
+	var anchorItems, nonAnchorItems []*IssuanceItem
```
Same here re pre-allocating the slice that we expect the majority of items to be in.

guggero reviewed (approved):
Awesome to see this all coming together, great work :100: 

Mostly nits and optimizations, so giving it the green checkmark :tada: 

Not sure if the coverage test failure is a flake, re-triggering it to see what's going on.

jharveyb commented on asset/asset.go:
```diff
@@ -428,7 +428,8 @@ type GroupKey struct {
 	// together across distinct asset IDs, allowing further issuance of the
 	// asset to be made possible. The tweaked public key is the result of:
-	//   groupInternalKey + sha256(groupInternalKey || genesisOutPoint) * G
+	// internalKey = rawKey + singleTweak * G
```
Didn't know about that feature before, nice.

halseth reviewed (approved):
This looks great now 👌 💯

jharveyb commented:
Rebased and addressed most recent feedback.

Roasbeef reviewed (approved):
LGTM 🍌
